tests/waku_relay/test_protocol.nim

```diff
@@ -0,0 +1,1303 @@
+      check:
+        anotherPeerManager.switch.isConnected(otherPeerId)
+        otherPeerManager.switch.isConnected(anotherPeerId)
+
+      # When publishing a message in anotherNode for each of the pubsub topics
+      handlerFuture = newPushHandlerFuture()
+      handlerFuture2 = newPushHandlerFuture()
+      otherHandlerFuture1 = newPushHandlerFuture()
+      otherHandlerFuture2 = newPushHandlerFuture()
+      anotherHandlerFuture1 = newPushHandlerFuture()
+      anotherHandlerFuture2 = newPushHandlerFuture()
+
+      let 
+        fromAnotherNodeWakuMessage4 = fakeWakuMessage("fromAnotherNode4")
+        fromAnotherNodeWakuMessage5 = fakeWakuMessage("fromAnotherNode5")
+        fromAnotherNodeWakuMessage6 = fakeWakuMessage("fromAnotherNode6")
+      
+      discard await anotherNode.publish(pubsubTopic, fromAnotherNodeWakuMessage4)
+      discard await anotherNode.publish(pubsubTopicB, fromAnotherNodeWakuMessage5)
+      discard await anotherNode.publish(pubsubTopicC, fromAnotherNodeWakuMessage6)
+
+      # Then the messages are published in all nodes except in node's B topic, because
+      # even if they're connected like so AnotherNode <-> OtherNode <-> Node,
+      # otherNode doesn't broadcast B topic messages because it's not subscribed to it
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        not await handlerFuture2.withTimeout(3.seconds)
+        await otherHandlerFuture1.withTimeout(3.seconds)
+        await otherHandlerFuture2.withTimeout(3.seconds)
+        await anotherHandlerFuture1.withTimeout(3.seconds)
+        await anotherHandlerFuture2.withTimeout(3.seconds)
+      
+      let
+        (topic7, msg7) = handlerFuture.read()
+        (otherTopic7, otherMsg7) = otherHandlerFuture1.read()
+        (otherTopic8, otherMsg8) = otherHandlerFuture2.read()
+        (anotherTopic7, anotherMsg7) = anotherHandlerFuture1.read()
+        (anotherTopic8, anotherMsg8) = anotherHandlerFuture2.read()
+      
+      check:
+        topic7 == pubsubTopic
+        msg7 == fromAnotherNodeWakuMessage4
+        otherTopic7 == pubsubTopic
+        otherMsg7 == fromAnotherNodeWakuMessage4
+        otherTopic8 == pubsubTopicC
+        otherMsg8 == fromAnotherNodeWakuMessage6
+        anotherTopic7 == pubsubTopicB
+        anotherMsg7 == fromAnotherNodeWakuMessage5
+        anotherTopic8 == pubsubTopicC
+        anotherMsg8 == fromAnotherNodeWakuMessage6
+
+      # Finally stop the other nodes
+      await allFutures(otherSwitch.stop(), otherNode.stop(), anotherSwitch.stop(), anotherNode.stop())
+
+  suite "Unsubscribe":
+    asyncTest "Without Subscription":
+      # Given an external topic handler
+      let 
+        otherSwitch = newTestSwitch()
+        otherNode = await newTestWakuRelay(otherSwitch)
+      await allFutures(otherSwitch.start(), otherNode.start())
+      let otherTopicHandler: TopicHandler = otherNode.subscribe(pubsubTopic, simpleFutureHandler)
```

Comment:
Why do you subscribe with `otherNode` and then checking `node` - what would be the potential side-effect of the `otherNode` being subscribed?

Reply:
To run an `unsubscribe` method I need a `TopicHandler`, which you get as return value from running `subscribe`. Thus, to test "Unsubscribe without subscription" I need a valid `TopicHandler` that the target node is unrelated to.

Reply:
Interesting:D I guess that is why I am not a QE as I would never think about testing this:D